tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -1165,6 +1188,151 @@ uint32_t Trace::instantiate(CommandQueue& cq) {
     return trace_id;
 }
 
+void EnqueueAddBufferToProgram(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer, std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>> program, bool blocking) {
+    cq.run_command(CommandInterface{
+        .type = EnqueueCommandType::ADD_BUFFER_TO_PROGRAM,
+        .blocking = blocking,
+        .buffer = buffer,
+        .program = program,
+    });
+}
+
+void EnqueueAddBufferToProgramImpl(const std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer, std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>> program) {
+    std::visit([program] (auto&& b) {
+        using buffer_type = std::decay_t<decltype(b)>;
+        if constexpr (std::is_same_v<buffer_type, std::shared_ptr<Buffer>>) {
+            std::visit([&b] (auto&& p) {
+                using program_type = std::decay_t<decltype(p)>;
+                if constexpr (std::is_same_v<program_type, std::reference_wrapper<Program>>) {
+                    p.get().add_buffer(b);
+                }
+                else {
+                    p->add_buffer(b);
+                }
+            }, program);
+        }
+    }, buffer);
+}
+
+void EnqueueUpdateRuntimeArgs(CommandQueue& cq, const std::shared_ptr<Kernel> kernel, const CoreCoord &core_coord, std::vector<uint32_t> &update_idx, std::shared_ptr<RuntimeArgs> runtime_args_vec, bool blocking) {
+    auto runtime_args_md = RuntimeArgsMetadata {
+            .core_coord = core_coord,
+            .runtime_args_vec = runtime_args_vec,
+            .kernel = kernel,
+            .update_idx = update_idx,
+    };
+    cq.run_command( CommandInterface {
+        .type = EnqueueCommandType::UPDATE_RUNTIME_ARGS,
+        .blocking = blocking,
+        .runtime_args_md = runtime_args_md,
+    });
+}
+
+void EnqueueUpdateRuntimeArgsImpl (const RuntimeArgsMetadata& runtime_args_md) {
+    std::vector<uint32_t> resolved_runtime_args = {};
+    resolved_runtime_args.reserve((*runtime_args_md.runtime_args_vec).size());
+
+    for (const auto& arg : *(runtime_args_md.runtime_args_vec)) {
```

Comment:
Weird that we call this attribute runtime_args_vec, but we're dereferencing it (so I'm guessing it's a pointer to a vector, not a vector).